File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

/**
 * Returns the result of a function applied to the first item in the given
 * array that satisfies the given predicate.
 *
 * @param items The array of items to search.
 * @param predicate The predicate to satisfy.
 * @param consumer The function to apply to the first item satisfying the predicate.
 * @returns The result of consumer applied to the first element in items that satisfies predicate. Undefined if no item
 * satisfies predicate.
 * */
export function firstThenApply<T, U>(
    items: T[],
    predicate: (item: T) => boolean,
    consumer: (item: T) => U
): U | undefined {
  for (const item of items) {
    if (predicate(item)) {
      return consumer(item)
    }
  }

  return undefined;
}

/**
 * Yields an infinite sequence of powers of a given base.
 *
 * @param base The base of which to generator powers.
 * @yields Sequential powers of the base (b^0, b^1, etc.).
 */
export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let power = 1n; ; power *= base) {
    yield power
  }
}

/**
 * Returns the number of lines in the given file that are (1) not empty, (2) not all whitespace, and (3) whose first
 * character is not '#'.
 *
 * @param filePath The path to the file whose lines will be counted. Throws an error if the path is not valid.
 * @returns The number of lines in the given file that satisfy specified requirements.
 */
export async function meaningfulLineCount(filePath: string): Promise<number> {
  let count = 0
  const file = await open(filePath, 'r')

  for await(const line of file.readLines()) {
    if ((line !== "") && (line.trim() !== "") && (line.trim()[0] !== "#")) {
      count++
    }
  }

  return count
}

interface Box {
  kind: "Box",
  width: number,
  length: number,
  depth: number
}

interface Sphere {
  kind: "Sphere",
  radius: number
}

export type Shape = Box | Sphere;

/**
 * Computes the surface area of the given shape.
 *
 * @param shape The shape whose surface area will be calculated (a box or a sphere).
 * @returns The surface area of the given shape.
 */
export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return 2 * ((shape.width * shape.length) + (shape.length * shape.depth) + (shape.depth * shape.width))
    case "Sphere":
      return 4 * Math.PI * (shape.radius ** 2)
  }
}

/**
 * Computers the volume of the given shape.
 *
 * @param shape The shape whose volume will be calculated (a box or a sphere).
 * @returns The volume of the given shape.
 */
export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return shape.width * shape.length * shape.depth
    case "Sphere":
      return (4 / 3) * Math.PI * (shape.radius ** 3)
  }
}

/** Interface representing a type of entry in a BST (Node or Empty). */
export interface BinarySearchTree<T> {
  /** Returns the total size of this tree (the number of non-empty nodes). */
  size(): number

  /**
   * Inserts a unique item into this BST.
   *
   * @param data The item to add to this tree.
   * @returns The node at which the item was inserted.
   */
  insert(data: T): BinarySearchTree<T>

  /** Returns whether this tree contains a specified value. */
  contains(data: T): boolean

  /** Yields each item in this tree in-order. */
  inorder(): Iterable<T>
}

/** An entry in a binary search tree with data. */
class Node<T> {
  /**
   * @param left This node's left child.
   * @param data The value contained in this node.
   * @param right This node's right child.
   */
  constructor(
      private readonly left: BinarySearchTree<T>,
      private readonly data: T,
      private readonly right: BinarySearchTree<T>
  ) {}

  size(): number {
    return this.left.size() + 1 + this.right.size()
  }

  insert(data: T): BinarySearchTree<T> {
    if (data < this.data) {
      return new Node<T>(this.left.insert(data), this.data, this.right)
    } else if (data > this.data) {
      return new Node<T>(this.left, this.data, this.right.insert(data))
    } else {
      return this
    }
  }
  contains(data: T): boolean {
    if (data == this.data) {
      return true
    } else if (data > this.data) {
      return this.right.contains(data)
    } else {
      return this.left.contains(data)
    }
  }
  *inorder(): Iterable<T> {
    yield* this.left.inorder()
    yield* [this.data]
    yield* this.right.inorder()
  }

  public toString(): string {
    return `(${this.left}${this.data}${this.right})`.replaceAll("()", "")
  }
}

/** An empty node in a binary search tree. */
export class Empty<T> {
  size(): number {
    return 0
  }

  insert(data: T): BinarySearchTree<T> {
    return new Node<T>(new Empty<T>(), data, new Empty<T>)
  }

  contains(data: T): boolean {
    return false
  }

  inorder(): Iterable<T> {
    return []
  }

  public toString(): string {
    return "()"
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [ 25; 10; 5; 1 ] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

(* Returns the application of a given function applied to the first item in the given list that satisfies the given
 * predicate. *)
let rec first_then_apply (item : 'a list) (pred : 'a -> bool) (func : 'a -> 'b option) : 'b option =
  match item with
  | [] -> None
  | x :: xs ->
  	if pred x then func x
    else first_then_apply xs pred func

(* Generates an infinite sequence of powers of a given base. *)
let powers_generator (base: int) : int Seq.t =
  let rec generate_power n () = Seq.Cons (n, generate_power (n * base)) in
  generate_power 1


(* Returns the number of lines in the given file that are (1) not empty, (2) not all whitespace, and (3) whose first
 * character is not '#'. *)
let meaningful_line_count (filename: string) : int =
  let is_meaningful_line (line: string) : bool =
    let trimmed_string = String.trim line in
    trimmed_string <> "" && not (String.starts_with ~prefix:"#" trimmed_string)
  in

  let file = open_in filename in
  let finally () = close_in file in

  let rec count_meaningful_lines (count: int) : int =
    match input_line file with
    | line -> count_meaningful_lines (if is_meaningful_line line then count + 1 else count)
    | exception End_of_file -> count
  in

  Fun.protect ~finally (fun () -> count_meaningful_lines 0)

(* Type representing a shape. Shapes implement methods for calculating their volume and surface area. *)
type shape = Sphere of float | Box of float * float * float

(* Calculates the volume of a given shape. *)
let volume (shape: shape) : float =
  match shape with
  | Sphere r -> Float.pi *. (r ** 3.) *. 4. /. 3.
  | Box (l, w, h) -> l *. w *. h

(* Calculates the surface area of a given shape. *)
let surface_area (shape: shape) : float =
  match shape with
  | Sphere r -> 4. *. Float.pi *. (r ** 2.)
  | Box (l, w, h) -> 2. *. ((l *. w) +. (l *. h) +. (w *. h))

(* Type representing an empty or non-empty node in a binary search tree. *)
type 'a binary_search_tree =
  | Empty
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

(* Returns the size of a binary search tree. *)
let rec size (tree: 'a binary_search_tree) : int =
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right

(* Returns whether a value is contained in a binary search tree. *)
let rec contains (value: 'a) (tree: 'a binary_search_tree) : bool =
  match tree with
  | Empty -> false
  | Node (left, v, right) ->
      if value = v then true
      else if value < v then contains value left
      else contains value right

(* Returns an in-order traversal of a binary search tree as a list. *)
let rec inorder (tree: 'a binary_search_tree) : 'a list=
  match tree with
  | Empty -> []
  | Node (left, v, right) -> inorder left @ [ v ] @ inorder right

(* Inserts a value into a binary search tree. *)
let rec insert (value: 'a) (tree: 'a binary_search_tree) : 'a binary_search_tree =
  match tree with
  | Empty -> Node (Empty, value, Empty)
  | Node (left, v, right) ->
      if value < v then Node (insert value left, v, right)
      else if value > v then Node (left, v, insert value right)
      else tree